### PR TITLE
Graphs : Bidirectional A*

### PR DESCRIPTION
**Pull request by guilyx**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: graphs/bidirectional_a_star.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        successors = []
+        for action in delta:
+            pos_x = parent.pos_x + action[1]
+            pos_y = parent.pos_y + action[0]
+            if (
+                pos_x < 0
+                or pos_x > len(grid[0]) - 1
+                or pos_y < 0
+                or pos_y > len(grid) - 1
+            ):
```
Comment (cclauss) [inline graphs/bidirectional_a_star.py:119]:
```suggestion
            if not (0 < pos_x < len(grid[0]) - 1 and 0 < pos_y < len(grid) - 1):
```
Impressive little optimization...  Compare the before / after times!

File: graphs/bidirectional_a_star.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+init = (0, 0)
+goal = (len(grid) - 1, len(grid[0]) - 1)
+for elem in grid:
+    print(elem)
+
+start_time = time.time()
+a_star = AStar(init, goal)
+a_star.search()
+end_time = time.time() - start_time
+print("AStar execution time = ", end_time, " seconds")
```
Comment (cclauss) [inline graphs/bidirectional_a_star.py:217]:
```suggestion
print(f"AStar execution time = {end_time:f} seconds")
```
Improve the ability to see very small floats.

File: graphs/bidirectional_a_star.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+bd_start_time = time.time()
+bidir_astar = BidirectionalAStar(init, goal)
+bidir_astar.search()
+bd_end_time = time.time() - bd_start_time
+print("BidirectionalAStar execution time = ", bd_end_time, " seconds")
```
Comment (cclauss) [inline graphs/bidirectional_a_star.py:223]:
```suggestion
print(f"BidirectionalAStar execution time = {bd_end_time:f} seconds")
```

File: graphs/bidirectional_a_star.py (hunk 2 above)
Comment (guilyx) [reply graphs/bidirectional_a_star.py:217]: Oh, didn't know about that one, thanks !

File: graphs/bidirectional_a_star.py (hunk 1 above)
Comment (guilyx) [reply graphs/bidirectional_a_star.py:119]: :+1: 

**Before** : 
```
AStar execution time =  0.22050976753234863  seconds
BidirectionalAStar execution time =  0.0417177677154541  seconds
```

**After** :
```
AStar execution time =  2.1457672119140625e-05  seconds
BidirectionalAStar execution time =  1.52587890625e-05  seconds
```

File: graphs/bidirectional_a_star.py (hunk 2 above)
Comment (guilyx) [reply graphs/bidirectional_a_star.py]: I do enjoy the "x.e-t" version though. Find it more readable than "0.00000x" to compare two values.

**Comment by deadshotsb** [comment]
good work @guilyx

**Pull request merged**